### DISCUSSION TITLE
Allow `out`, `in`, and `ref` in generic type parameters.

### DISCUSSION THREAD

**Discussion by colejohnson66** [dotnet/csharplang#7839] [Language Ideas] [answered]
Currently, generics don't support `out`, `in`, or `ref` on type parameters. This causes issues when one wants to use `Action` and `Func` delegates. For example, `Func<ref int, int>` does not work. Instead, the programmer is forced to create their own delegate type:

```cs
public delegate TOutput RefFunc<TInput, TOutput>(ref TInput param);
```

I propose that this limitation be lifted so that one could write `Func<ref int, int>` without error. Normal generic methods would work the same way:

```cs
int x = 0;
Test<ref int>(ref x);

void Test<T>(T param) { }
```

The compiler would "fill in" the type parameters into the definition, and complain if something doesn't make sense. For example, the following would cause a compiler error:

```cs
Func<out int> x; // error about usage of an "out" parameter in return position
Test<in int>();  // error about usage of an "in" parameter in return position

T Test<T>() => default(T)!;
```

**Comment by HaloFour** [comment]
You can just define a different delegate type to support `ref` or `out` parameters.  The runtime forbids the use of `ref` types for generic parameters as the IL required to work with them are necessarily different.

**Comment by siegfriedpammer** [comment, marked as answer]
There is a proposal for this already #5321

**Comment by CyrusNajmabadi** [comment]
We'd likely do this, if the runtime adds support for it.

**Reply by HaloFour** [reply]
From the proposal it sounds like it would be limited to delegates, and only where the use is defined as `ref`, so you couldn't use `ref T` for `M<T>(Action<T>)`, but you could use `M<T>(Action<ref T>)`.  Sounds like a lot of hoops for very little benefit.
